src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs

```diff
@@ -66,70 +65,183 @@ public void WriteMessage(HubMessage message, Stream output)
+                                            result = PayloadSerializer.Deserialize(reader);
+                                        }
+                                    }
+                                    else if (string.Equals(memberName, ItemPropertyName, StringComparison.Ordinal))
+                                    {
+                                        reader.Read();
+
+                                        if (string.IsNullOrEmpty(invocationId))
+                                        {
+                                            // If we don't have an invocation id then we need to store it as a JToken so we can parse it later
+                                            itemToken = JToken.Load(reader);
+                                        }
+                                        else
+                                        {
+                                            var returnType = binder.GetReturnType(invocationId);
+                                            item = PayloadSerializer.Deserialize(reader);
+                                        }
+                                    }
+                                    else if (string.Equals(memberName, ArgumentsPropertyName, StringComparison.Ordinal))
+                                    {
+                                        reader.Read();
+
+                                        JsonUtils.EnsureTokenType(ArgumentsPropertyName, reader.TokenType, JsonToken.StartArray);
+
+                                        if (string.IsNullOrEmpty(target))
+                                        {
+                                            // We don't know the method name yet so just parse an array of generic JArray
+                                            argumentToken = JArray.Load(reader);
+                                        }
+                                        else
+                                        {
+                                            try
+                                            {
+                                                var paramTypes = binder.GetParameterTypes(target);
+                                                arguments = BindArguments(reader, paramTypes);
+                                            }
+                                            catch (Exception ex)
+                                            {
+                                                argumentBindingException = ExceptionDispatchInfo.Capture(ex);
+                                            }
+                                        }
+                                    }
+                                    else if (string.Equals(memberName, HeadersPropertyName, StringComparison.Ordinal))
+                                    {
+                                        reader.Read();
+                                        headers = _jsonSerializer.Deserialize<Dictionary<string, string>>(reader);
```

Comment:
This will be faster manually read with `JsonReader`.

Reply:
I bet but headers aren't in our main scenario and I didn't want to think about it. I guess it isn't that bad since we're just parsing string ->string. Still I don't think it's worth it.

Reply:
A sample I prepared earlier https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Converters/ExpandoObjectConverter.cs#L113-L141

Change the dictionary type line 115 and the change the `ReadValue` line 129 to a check that the token type is a string and then set the `reader.Value` into the dictionary.

Reply:
Will do.